### PR TITLE
gRPC: fail code generation gracefully on default java package

### LINKED ISSUE #17485: gRPC: Mutiny classes generation fails without java package definition

### PR DESCRIPTION
**Pull request by michalszynkiewicz**
fixes #17485

### CODE REVIEW

**Comment by michalszynkiewicz** [comment]
For the user, the error looks as follows:
```java
Using default java package is not supported for gRPC code generation. Please specify `option java_package = "your.package"` in my-new.proto
java.lang.IllegalArgumentException: Code generation failed. Please check the log above for details.
        at io.quarkus.grpc.protoc.plugin.MutinyGrpcGenerator.validateServices(MutinyGrpcGenerator.java:59)
        at io.quarkus.grpc.protoc.plugin.MutinyGrpcGenerator.generateFiles(MutinyGrpcGenerator.java:44)
        at com.salesforce.jprotoc.ProtocPlugin.lambda$generate$1(ProtocPlugin.java:191)
        at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:271)
        at java.base/java.util.Collections$2.tryAdvance(Collections.java:4745)
        at java.base/java.util.Collections$2.forEachRemaining(Collections.java:4753)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.StreamSpliterators$WrappingSpliterator.forEachRemaining(StreamSpliterators.java:312)
        at java.base/java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:735)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
        at com.salesforce.jprotoc.ProtocPlugin.generate(ProtocPlugin.java:196)
        at com.salesforce.jprotoc.ProtocPlugin.generate(ProtocPlugin.java:80)
        at com.salesforce.jprotoc.ProtocPlugin.generate(ProtocPlugin.java:51)
        at com.salesforce.jprotoc.ProtocPlugin.generate(ProtocPlugin.java:43)
        at io.quarkus.grpc.protoc.plugin.MutinyGrpcGenerator.main(MutinyGrpcGenerator.java:311)
--q-grpc_out: protoc-gen-q-grpc: Plugin failed with status code 1.

```

File: extensions/grpc/protoc/src/main/java/io/quarkus/grpc/protoc/plugin/MutinyGrpcGenerator.java [hunk 1]
```diff
@@ -41,9 +41,25 @@ private String getMethodJavaDocPrefix() {
...
+        validateServices(services);
         return generateFiles(services);
     }
 
+    private void validateServices(List<ServiceContext> services) {
+        boolean failed = false;
+        for (ServiceContext service : services) {
+            if (service.packageName == null || service.packageName.isBlank()) {
+                System.err.println("Using the default java package is not supported for "
+                        + "gRPC code generation. Please specify `option java_package = \"your.package\"` in "
```
Comment (cescoffier) [inline extensions/grpc/protoc/src/main/java/io/quarkus/grpc/protoc/plugin/MutinyGrpcGenerator.java:53]: gRPC :-D

Comment (cescoffier) [reply extensions/grpc/protoc/src/main/java/io/quarkus/grpc/protoc/plugin/MutinyGrpcGenerator.java:53]: I would also use a logger and the "error" level.

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: extensions/grpc/protoc/src/main/java/io/quarkus/grpc/protoc/plugin/MutinyGrpcGenerator.java (hunk 1 above)
Comment (michalszynkiewicz) [reply extensions/grpc/protoc/src/main/java/io/quarkus/grpc/protoc/plugin/MutinyGrpcGenerator.java:53]: do you mean it should be something like `Quarkus gRPC code generation`?

Comment (michalszynkiewicz) [reply extensions/grpc/protoc/src/main/java/io/quarkus/grpc/protoc/plugin/MutinyGrpcGenerator.java:53]: for logging, only java util logging is available there, I can switch to that. Another option is to gather all the problems and print them in the exception's message

Comment (cescoffier) [reply extensions/grpc/protoc/src/main/java/io/quarkus/grpc/protoc/plugin/MutinyGrpcGenerator.java]: oh right, you are in the code generation phase. Ok, let's keep System.out for now. We may want to add some log support in the future.

File: extensions/grpc/protoc/src/main/java/io/quarkus/grpc/protoc/plugin/MutinyGrpcGenerator.java [hunk 2]
```diff
@@ -19,6 +21,8 @@
  */
 public class MutinyGrpcGenerator extends Generator {
 
+    private static final Logger log = Logger.getLogger(MutinyGrpcGenerator.class.getName());
```
Comment (cescoffier) [inline extensions/grpc/protoc/src/main/java/io/quarkus/grpc/protoc/plugin/MutinyGrpcGenerator.java:24]: +1

File: extensions/grpc/protoc/src/main/java/io/quarkus/grpc/protoc/plugin/MutinyGrpcGenerator.java [hunk 3]
```diff
@@ -41,9 +45,25 @@ private String getMethodJavaDocPrefix() {
...
+        validateServices(services);
         return generateFiles(services);
     }
 
+    private void validateServices(List<ServiceContext> services) {
+        boolean failed = false;
+        for (ServiceContext service : services) {
+            if (service.packageName == null || service.packageName.isBlank()) {
+                log.log(Level.SEVERE, "Using the default java package is not supported for "
+                        + "Quarkus RPC code generation. Please specify `option java_package = \"your.package\"` in "
```
Comment (cescoffier) [inline extensions/grpc/protoc/src/main/java/io/quarkus/grpc/protoc/plugin/MutinyGrpcGenerator.java:57]:
```suggestion
                        + "Quarkus gRPC code generation. Please specify `option java_package = \"your.package\"` in "
```

**Review by cescoffier — CHANGES_REQUESTED** [review]
Just a typo in the message.

File: extensions/grpc/protoc/src/main/java/io/quarkus/grpc/protoc/plugin/MutinyGrpcGenerator.java (hunk 3 above)
Comment (michalszynkiewicz) [reply extensions/grpc/protoc/src/main/java/io/quarkus/grpc/protoc/plugin/MutinyGrpcGenerator.java]: fixed

File: extensions/grpc/protoc/src/main/java/io/quarkus/grpc/protoc/plugin/MutinyGrpcGenerator.java (hunk 1 above)
Comment (michalszynkiewicz) [reply extensions/grpc/protoc/src/main/java/io/quarkus/grpc/protoc/plugin/MutinyGrpcGenerator.java]: changed to JUL

**Review by cescoffier — APPROVED** [review]

**Pull request merged**